cpp/arcticdb/storage/lmdb/lmdb_mock_client.cpp

```diff
@@ -0,0 +1,135 @@
+/* Copyright 2024 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/storage/lmdb/lmdb_mock_client.hpp>
+
+#include <arcticdb/codec/segment.hpp>
+#include <arcticdb/entity/atom_key.hpp>
+#include <arcticdb/entity/serialized_key.hpp>
+#include <arcticdb/util/string_utils.hpp>
+
+
+namespace arcticdb::storage::lmdb {
+
+std::string MockLmdbClient::get_failure_trigger(
+        const std::string& path,
+        StorageOperation operation_to_fail,
+        int error_code) {
+    return fmt::format("{}#Failure_{}_{}", path, operation_to_string(operation_to_fail), error_code);
+}
+
+std::string lmdb_operation_string(StorageOperation operation) {
+    switch (operation) {
+        case StorageOperation::READ:
+            return "mdb_get";
+        case StorageOperation::WRITE:
+            return "mdb_put";
+        case StorageOperation::DELETE:
+            return "mdb_del";
+        case StorageOperation::LIST:
+            return "mdb_cursor_get";
+        case StorageOperation::EXISTS:
+            return "mdb_get";
+        default:
+            return "unknown";
+    }
+}
+
+void raise_if_has_failure_trigger(const LmdbKey& key, StorageOperation operation) {
+    auto path = key.path;
+    auto failure_string_for_operation = "#Failure_" + operation_to_string(operation) + "_";
+    auto position = path.rfind(failure_string_for_operation);
+
+    if (position == std::string::npos) {
+        return;
+    }
+
+    int error_code = 0;
+    try {
+        auto start = position + failure_string_for_operation.size();
+        error_code = stoi(path.substr(start));
+        auto error_message = fmt::format("Simulated Error, message: operation {}, error code {}",
+                                         operation_to_string(operation), error_code);
+    } catch (std::exception&) {
+        return;
+    }
+
+    if (error_code != 0) {
+        ::lmdb::error::raise(lmdb_operation_string(operation).data(), error_code);
+    }
+}
+
+void raise_key_exists_error(std::string lmdb_op) {
+    ::lmdb::error::raise(lmdb_op.data(), MDB_KEYEXIST);
+}
+
+bool MockLmdbClient::has_key(const LmdbKey& key) const {
+    return lmdb_contents.find(key) != lmdb_contents.end();
+}
+
+bool MockLmdbClient::exists(const std::string& db_name, std::string& path, ::lmdb::txn&, ::lmdb::dbi&) const {
```

Comment:
Why is path a ref do we expect the function to change it somehow?

Reply:
In `RealLmdbClient::exists` `path` is used to initialize `MDB_val mdb_key` which doesn't accept const. Therefore, it's not const there. Both `RealLmdbClient::exists` and `MockLmdbClient::exists` definitions must match as they extend `LmdbClientWrapper`.